File: backend/app/api/calendar.py
from datetime import date

from fastapi import APIRouter, Depends, Query
from fastapi.responses import Response
from sqlalchemy.orm import Session

from app.api.deps import get_current_user
from app.core.database import get_db
from app.models.user import User
from app.schemas.calendar_event import (
    CalendarEventCreate,
    CalendarEventOut,
    CalendarEventUpdate,
    DaySnapshotOut,
    DaySnapshotProductionOut,
    DaySnapshotSaleOut,
    MentionableUserOut,
)
from app.services import calendar_service

router = APIRouter(prefix="/api/calendar", tags=["calendar"])


@router.get("/events", response_model=list[CalendarEventOut])
def list_events(
    year: int = Query(..., ge=1970, le=2100),
    month: int = Query(..., ge=1, le=12),
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    events = calendar_service.list_events_for_month(db, user, year, month)
    return [CalendarEventOut.from_model(e, user.id) for e in events]


@router.get("/events.ics")
def export_ics(
    year: int = Query(..., ge=1970, le=2100),
    month: int = Query(..., ge=1, le=12),
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    events = calendar_service.list_events_for_month(db, user, year, month)
    ics_text = calendar_service.build_ics(events)
    return Response(
        content=ics_text,
        media_type="text/calendar",
        headers={"Content-Disposition": f"attachment; filename=calendar-{year:04d}-{month:02d}.ics"},
    )


@router.get("/mentionable-users", response_model=list[MentionableUserOut])
def mentionable_users(db: Session = Depends(get_db), _: User = Depends(get_current_user)):
    return calendar_service.list_mentionable_users(db)


@router.get("/day-snapshot", response_model=DaySnapshotOut)
def day_snapshot(
    target_date: date = Query(..., alias="date"),
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
):
    """Powers the calendar's day-actions popup: what's already logged
    for this day (production batches, sales orders) plus whether it's
    still a legal target for logging something new."""
    snapshot = calendar_service.get_day_snapshot(db, target_date)
    return DaySnapshotOut(
        date=snapshot["date"],
        production=[DaySnapshotProductionOut.from_model(b) for b in snapshot["production"]],
        sales=[DaySnapshotSaleOut.from_model(o) for o in snapshot["sales"]],
        can_log=snapshot["can_log"],
    )


@router.post("/events", response_model=CalendarEventOut, status_code=201)
def create_event(
    payload: CalendarEventCreate,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    event = calendar_service.create_event(db, user, payload)
    return CalendarEventOut.from_model(event, user.id)


@router.put("/events/{event_id}", response_model=CalendarEventOut)
def update_event(
    event_id: int,
    payload: CalendarEventUpdate,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    event = calendar_service.update_event(db, user, event_id, payload)
    return CalendarEventOut.from_model(event, user.id)


@router.delete("/events/{event_id}", status_code=204)
def delete_event(
    event_id: int,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    calendar_service.delete_event(db, user, event_id)

File: frontend/src/api/calendar.ts
import { apiClient } from './client'
import type { CalendarEvent, CalendarEventInput, DaySnapshot, MentionableUser } from '@/types/calendar'

export async function listCalendarEvents(year: number, month: number): Promise<CalendarEvent[]> {
  const { data } = await apiClient.get<CalendarEvent[]>('/api/calendar/events', { params: { year, month } })
  return data
}

export async function createCalendarEvent(payload: CalendarEventInput): Promise<CalendarEvent> {
  const { data } = await apiClient.post<CalendarEvent>('/api/calendar/events', payload)
  return data
}

export async function updateCalendarEvent(id: number, payload: CalendarEventInput): Promise<CalendarEvent> {
  const { data } = await apiClient.put<CalendarEvent>(`/api/calendar/events/${id}`, payload)
  return data
}

export async function deleteCalendarEvent(id: number): Promise<void> {
  await apiClient.delete(`/api/calendar/events/${id}`)
}

export async function listMentionableUsers(): Promise<MentionableUser[]> {
  const { data } = await apiClient.get<MentionableUser[]>('/api/calendar/mentionable-users')
  return data
}

/** What's already logged for a given day (production, sales) plus
 * whether it's still a legal target for logging something new -- powers
 * the calendar's day-actions popup. */
export async function getDaySnapshot(isoDate: string): Promise<DaySnapshot> {
  const { data } = await apiClient.get<DaySnapshot>('/api/calendar/day-snapshot', { params: { date: isoDate } })
  return data
}

/** Downloads the ICS-compatible export for a given month and triggers a
 * browser save -- goes through apiClient (not a plain <a href>) since
 * the endpoint requires the bearer token the same as every other call. */
export async function downloadCalendarIcs(year: number, month: number): Promise<void> {
  const response = await apiClient.get('/api/calendar/events.ics', {
    params: { year, month },
    responseType: 'blob',
  })
  const url = URL.createObjectURL(response.data as Blob)
  const link = document.createElement('a')
  link.href = url
  link.download = `calendar-${year}-${String(month).padStart(2, '0')}.ics`
  document.body.appendChild(link)
  link.click()
  link.remove()
  URL.revokeObjectURL(url)
}
